src/test/clojure/xtdb/pgwire_protocol_test.clj

```diff
@@ -0,0 +1,345 @@
+  (let [buffer (ByteBuffer/allocate 4)]
+    (.putInt buffer value)
+    (.array buffer)))
+
+(defrecord WrappedSocket [^Socket socket in out]
+  AutoCloseable
+  (close [_] (.close socket)))
+
+(defn wrap-socket [^Socket socket]
+  (let [in (DataInputStream. (.getInputStream socket))
+        out (DataOutputStream. (.getOutputStream socket))]
+    (->WrappedSocket socket in out)))
+
+(defn connect ^AutoCloseable [^long port]
+  (wrap-socket (Socket. "localhost" port)))
+
+(defn write-string-parameter [^DataOutputStream out ^String key ^String value]
+  (doto out
+    (.writeBytes key)
+    (.write 0)
+    (.writeBytes value)
+    (.write 0)))
+
+(defn bytes->str [^bytes arr]
+  (String. arr StandardCharsets/UTF_8))
+
+(defn read-message [^DataInputStream in]
+  (let [response-type (.read in)]
+    (when (= response-type -1)
+      (throw (IOException. "Unexpected end of stream")))
+    (let [response-char (char response-type)
+          _length (.readInt in)]
+
+      (if-let [{:keys [read name]} @(get pgwire/server-msgs response-char)]
+
+        (let [{auth-code :result :as message} (-> (read in)
+                                                  (assoc :message-type name))]
+          (case response-char
+            ;; 'R' indicates an Authentication request
+            \R
+            (cond
+              (= auth-code 0)
+              message
+
+              :else ;; TODO add other authentication methods if available
+              (throw (IOException. (str "Unsupported authentication method: " auth-code))))
+
+
+            ;; 'D' indicates a DataRow message (contains a row of query result data)
+            ;; we are not actually employing the send decoding types here, but string is more readable
+            \D (update message :vals (partial mapv bytes->str))
+
+            message))
+
+        (throw (IllegalStateException. (str "Unhandled message type: " response-char)))))))
+
+(defn read-response
+  ([^InputStream in]
+   (loop [responses []]
+     (let [{:keys [message-type] :as  response} (read-message in)]
+       (log/info response)
+       (if (= message-type "msg-ready")
+         (conj responses response)
+         (recur (cond-> responses response (conj response))))))))
+
+(defn startup [{:keys [in ^DataOutputStream out]} user database]
+  (let [startup-message-ba (ByteArrayOutputStream.)
+        startup-message (DataOutputStream. startup-message-ba)]
+    ;; Reserve 4 bytes for the message length (to be set later)
+    (.writeInt startup-message 0)
+    ;; this is the normal 'hey I'm a postgres client' if ssl is not requested
+    (.writeInt startup-message 196608)
+    (write-string-parameter startup-message "user" user)
+    (write-string-parameter startup-message "database" database)
+    (.write startup-message 0)
+    (let [^bytes message-bytes (.toByteArray startup-message-ba)
+          length (alength message-bytes)]
+      (System/arraycopy (int-to-bytes length) 0 message-bytes 0 4)
```

Comment:
I tried to rewrite this function in the style of all the others, but something goes awefully wrong in `reac-c-string` on the server side. I haven't fully figured out what is wrong. It is probably good to debug it, but didn't want to blog this PR on it.